피드 기능 구현

YehyeokBang opened the pull request:
## 내용

피드, 댓글, 좋아요 기능을 구현합니다.

- 피드 작성 및 조회(목록, 단일)
- 댓글 작성 기능 (1depth)
- 좋아요 및 취소 기능
- 관련 테스트 코드 작성

> 임시로 구현한 부분이 많아서 사용해 보면서 수정이 필요합니다.

## 필요한 기능

- 조회수 기능 (현재는 LMS 처럼 조회하면 제한 없이 조회수가 늘어납니다.)
- 사용자 엔티티 추출 기능 (반복되는 코드를 줄일 필요가 있을 것 같습니다.)
- 각 피드에 사용자의 위치 정보(동네 정보)는 미구현 상태입니다. 사용자 동네 정보 기능부터 구현 후 추가할 예정입니다.

## 좋아요

- 좋아요 취소(자신이 해당 피드에 좋아요를 눌렀는가?)를 위한 복합 인덱스(사용자 id, 피드 id) 추가

아래는 두 실행 계획을 표 형식으로 정리한 비교입니다.

| **항목**                  | **복합 인덱스 사용 실행 계획**                                                     | **인덱스 미설정 실행 계획**                                                 |
|---------------------------|---------------------------------------------------------------------------|-----------------------------------------------------------------------|
| **인덱스 조건**            | `USER_ID`와 `FEED_ID`를 모두 활용                                        | `USER_ID`만 활용, `FEED_ID`는 추가적으로 필터링 필요                 |
| **scanCount**             | 1 (한 번의 스캔으로 결과 탐색)                                           | 2 (인덱스 스캔 후 테이블에서 추가 필터링)                            |
| **효율성**                 | 더 효율적, 조건에 맞는 결과를 바로 반환                                  | 덜 효율적, 필터링 작업으로 인해 추가 연산 발생                        |
| **쿼리 성능 기대치**       | 높은 성능, 대상 데이터를 빠르게 검색 가능                                | 상대적으로 낮은 성능, 조건을 만족하는 데이터에 대해 추가 검사가 필요 |
| **적합한 경우**            | `USER_ID`와 `FEED_ID`가 자주 결합되어 조건으로 사용될 경우 적합          | `USER_ID` 조건이 주로 사용되고, `FEED_ID`가 부차적인 경우 적합        |

---

### **분석 요약**
1. **`scanCount` 차이**:
   - 첫 번째 쿼리는 복합 인덱스를 사용하여 조건을 한 번의 스캔으로 처리했습니다.  
   - 두 번째 쿼리는 단일 인덱스에서 검색한 뒤, 결과에서 `FEED_ID`를 추가로 필터링하는 작업이 수행되었습니다.

YehyeokBang committed:
[#22] feat: 피드 및 댓글 엔티티 작성

- 피드 작성 및 조회 기능을 구현합니다.
- 조회수 로직은 미구현 상태입니다.

YehyeokBang committed:
[#22] feat: 댓글 작성 및 삭제 기능 구현

댓글 작성자만 댓글을 삭제할 수 있습니다.

YehyeokBang committed:
[#22] refactor: 피드 목록 조회 기능 정렬

작성 시간 내림차순으로 정렬합니다.

YehyeokBang committed:
[#22] feat: 피드 좋아요 및 취소 기능 구현

각 사용자는 하나의 피드에 하나의 좋아요만 누를 수 있습니다.

YehyeokBang committed:
refactor: 테스트 코드 DisplayName 위치 변경

YehyeokBang committed:
[#22] test: 피드 관련 테스트 코드 작성

YehyeokBang committed:
[#22] test: 피드 댓글 관련 테스트 코드 작성

YehyeokBang committed:
[#22] refactor: 불필요한 static 키워드 제거

YehyeokBang committed:
[#22] test: 좋아요 관련 테스트 코드 작성

sss4920 commented on Network-Office/src/main/java/dev/office/networkoffice/feed/entity/Feed.java:
```diff
@@ -0,0 +1,68 @@
+    private String category;
+
+    @Column(name = "views", nullable = false)
+    private Long view = 0L;
```
아예 null값을 false로 만든다면 그냥 long 타입을 사용하는 것은 어떤지 궁금합니다!

sss4920 commented on Network-Office/src/main/java/dev/office/networkoffice/feed/entity/Likes.java:
```diff
@@ -0,0 +1,51 @@
+    private User user;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "feed_id")
```
좋아요에서 피드는 null인 경우가 있는 걸까용??

YehyeokBang replied:
피드가 삭제된 경우 어떻게 처리할지에 대해 고민하다가, 현재는 별도의 조건을 추가하지 않았습니다. 
허나 지금 생각해 보니 피드 삭제 시 좋아요를 자동으로 함께 삭제하도록 하는 것이 직관적일 것 같다는 생각이 드네요. JPA의 `cascade` 속성을 활용해보겠습니다! 👍

sss4920 commented on Network-Office/src/main/java/dev/office/networkoffice/feed/entity/Likes.java:
```diff
@@ -0,0 +1,51 @@
+        return new Likes(user, feed);
+    }
+
+    private static void validateArguments(User user, Feed feed) {
```
오 이런식으로 유효성 검사도 따로 빼주는거 보기 편한거같아요!☺️

sss4920 commented on Network-Office/src/main/java/dev/office/networkoffice/feed/service/CommentService.java:
```diff
@@ -0,0 +1,63 @@
+    public void removeComment(Long userId, Long commentId) {
+        User author = findUserById(userId);
+        Comment targetComment = findCommentById(commentId);
+        checkAuthor(targetComment, author);
```
`checkAuthor`라는 함수는 댓글 안에서 저자가 맞는지 검사하는게 더 엔티티의 역할에 맞게 구현할 수 있지않을까하는 생각이 듭니다! 만약 그렇게 된다면 다음과 같이 적을 수 있을 것 같아요..!
```java
if(targetComment.checkAuthor()){
  commentRepository.delete(targetComment);
}
```

YehyeokBang replied:
좋은 방법인 것 같아요! 적용하겠습니다!

sss4920 commented on Network-Office/src/main/java/dev/office/networkoffice/feed/service/CommentService.java:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    private boolean isAuthor(Comment targetComment, User author) {
```
isAuthor 역시 Comment안에서 해결해주면 서비스 로직이 조금 더 서비스에만 집중할 수 있게 될 것 같습니다!!

sss4920 commented on Network-Office/src/test/java/dev/office/networkoffice/feed/entity/LikesTest.java:
```diff
@@ -0,0 +1,65 @@
+
+    @DisplayName("User가 null인 경우 Likes 생성 시 예외 발생")
+    @ParameterizedTest
+    @CsvSource({
```
오 카테고리쪽에서와 다르게 여기서는 `null`로 주신 이유가 있을까용??

YehyeokBang replied:
두 케이스 모두 포함시킨다면 더 좋을 것 같습니다! 후반에 따로 좋아요를 분리하는 작업을 진행하면서 놓친 것 같아요..

sss4920 reviewed (changes requested):
수고하셨습니다!! 최근에 프리코스 활동으로 엄청 바쁘셨던 것 같은데 테스트 코드까지 잘짜신 것 같네요..! 몇 가지 질문 사항을 적어봤습니다! ☺️

YehyeokBang committed:
feat: id가 null일 경우를 위한 equals 로직 추가

- id가 null일 경우 소셜 계정 정보를 확인합니다.

YehyeokBang committed:
[#22] refactor: 작성자 확인 책임을 엔티티로 변경

- 댓글이 스스로 작성자를 판단할 수 있도록 변경합니다.
- 관련 테스트 코드를 추가로 작성합니다.

YehyeokBang committed:
[#22] fix: 요청 본문 오류 수정

- RequestBody 어노테이션을 추가합니다.

YehyeokBang committed:
[#22] refactor: 좋아요 기능 쿼리 최적화

- 좋아요 기능 쿼리 최적화 및 복합 인덱스를 설정합니다.
- 불필요한 로직을 제거합니다.

YehyeokBang committed:
refactor: 불필요한 Wrapper 클래스 제거

- 논리적으로 null이 포함될 수 없기에 Wrapper 클래스 대신 기본 타입을 사용하는 것이 더 적합하다고 판단하여 조회수 필드를 Long에서 long으로 변경합니다.

YehyeokBang commented:
- 저자 검증 기능을 엔티티 객체가 스스로 할 수 있도록 변경합니다.
- 논리적으로 null이 들어갈 수 없는 필드를 Wrapper 클래스에서 기본 타입으로 변경합니다.
- 좋아요 쿼리 개선 및 복합 인덱스를 설정합니다.
- 피드가 삭제된 경우 좋아요도 삭제될 수 있도록 cascade 설정을 추가합니다.
- 관련 테스트 코드를 추가합니다.

@sss4920 @djdongjae 리뷰 반영했습니다!

sss4920 commented on Network-Office/src/main/java/dev/office/networkoffice/feed/service/CommentService.java:
```diff
@@ -51,13 +51,8 @@ private Comment findCommentById(Long commentId) {
     private void checkAuthor(Comment targetComment, User author) {
-        if (!isAuthor(targetComment, author)) {
-            throw new IllegalArgumentException("댓글 작성자가 아닌 경우 댓글을 삭제할 수 없습니다.");
+        if (!targetComment.isCreatedBy(author)) {
```
오 읽기 너무 좋은 것 같습니다..!☺️

sss4920 reviewed (approved):
수고하셨습니다!☺️

djdongjae commented on Network-Office/src/main/java/dev/office/networkoffice/feed/entity/Comment.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public boolean isCreatedBy(User targetAuthor) {
+        return author.equals(targetAuthor);
```
`NullPointerException` 방지를 위해 targetAuthor가 null일 경우를 대비하는 것은 어떨까요?

YehyeokBang replied:
좋은 방법 같아요! 👍

djdongjae commented on Network-Office/src/main/java/dev/office/networkoffice/feed/service/FeedService.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Transactional(readOnly = true)
+    public Slice<FeedInfo> getFeeds(Pageable pageable) {
+        return feedRepository.findAllByOrderByCreatedTimeDesc(pageable)
```
`pageable` 객체도 null일 가능성이 있을 것 같습니다!

YehyeokBang replied:
Default 값을 지정하면 좋을 것 같네요 감사합니다!

djdongjae reviewed (approved):
역시 메소드별로 책임을 분리해서 코드를 잘 짜시는 것 같습니다..! 고생하셨습니다 🤗

YehyeokBang committed:
[#22] feat: 비교 대상 사용자 객체가 null인 경우 처리 구현

YehyeokBang committed:
[#22] feat: 페이지네이션 기본값 설정

- Swagger 문서도 기본값을 지정합니다.
